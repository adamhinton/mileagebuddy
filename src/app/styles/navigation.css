/* Styling for Header and its related components, including Tabs.tsx */

@tailwind components;

@layer components {
	/* Navigation link styling */
	.nav-link {
		@apply px-3 py-2 rounded-md text-neutral-text hover:bg-background-highlight 
    transition-colors duration-200 text-sm font-medium relative
    after:absolute after:bottom-0 after:left-0 after:h-0.5 after:w-0 after:bg-primary
    hover:after:w-full after:transition-all;
	}

	.nav-link-primary {
		@apply px-4 py-2 rounded-md  hover:bg-opacity-90 
    text-white text-sm font-medium transition-all shadow-sm 
    hover:shadow dark:bg-primary-200 dark:hover:bg-primary-100;
	}

	.nav-link-outline {
		@apply px-4 py-2 rounded-md border border-primary text-primary
    hover:bg-primary hover:bg-opacity-10 text-sm font-medium transition-all
    dark:border-primary-200 dark:text-primary-100 dark:hover:bg-primary-50 
    dark:hover:bg-opacity-20 dark:bg-background-elevated;
	}

	/* Tabs component styles */
	.mileage-calc-tabs-container {
		@apply md:sticky md:top-0 md:z-30 bg-background-header shadow-md;
	}

	.mileage-calc-tabs-inner {
		@apply container mx-auto px-2 sm:px-4;
	}

	.mileage-calc-tabs-list {
		@apply flex flex-wrap text-xs sm:text-sm font-medium text-center text-neutral-text border-b-2 border-primary-50 dark:border-primary-100/30;
	}

	.mileage-calc-tabs-item {
		@apply me-1 sm:me-2;
	}

	.mileage-calc-tabs-link {
		@apply inline-block p-2 sm:p-3 md:p-4 rounded-t-lg;
	}

	.mileage-calc-tabs-link--active {
		@apply bg-primary-50 text-primary font-semibold border-b-2 border-primary dark:bg-primary-100/20 dark:border-primary-200;
	}

	.mileage-calc-tabs-link--inactive {
		@apply hover:bg-background-highlight hover:text-primary transition-colors;
	}
}
