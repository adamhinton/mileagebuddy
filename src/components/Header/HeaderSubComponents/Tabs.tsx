"use client";

import Link from "next/link";

// ________________________________________________________________________
// This is tabs that handle internal navigation
// Has links to all internal pages
// Thanks to Flowbite, whom I shamelessly ripped this off from:
// https://flowbite.com/docs/components/tabs/
// This is toggleable. It's also collapsed by default on

// TODO tabs:
// -- Delineate this visually from Header.tsx
// -- Organize styling
// -- Unit tests
// -- Make it know which tab it's on (active tab)
// -- Don't refresh or do anything when clicking currently active tab
// Toggleable on mobile
// ________________________________________________________________________

const Tabs = () => {
	return (
		<section className="bg-background-elevated py-2 shadow-sm">
			<div className="container mx-auto px-4">
				<ul className="flex flex-wrap text-sm font-medium text-center text-neutral-text border-b border-primary-50 dark:border-primary-100/30">
					<li className="me-2">
						<Link
							href="/login"
							aria-current="page"
							className="inline-block p-3 bg-primary-50 text-primary rounded-t-lg border-b-2 border-primary dark:bg-primary-100/20 dark:border-primary-200"
						>
							Login
						</Link>
					</li>
					<li className="me-2">
						<Link
							href="/dashboard"
							className="inline-block p-3 rounded-t-lg hover:bg-background-highlight hover:text-primary transition-colors"
						>
							Dashboard
						</Link>
					</li>
					<li className="me-2">
						<Link
							href="/calculator"
							className="inline-block p-3 rounded-t-lg hover:bg-background-highlight hover:text-primary transition-colors"
						>
							Calculator
						</Link>
					</li>
					<li className="me-2">
						<Link
							href="/"
							className="inline-block p-3 rounded-t-lg hover:bg-background-highlight hover:text-primary transition-colors"
						>
							About
						</Link>
					</li>
				</ul>
			</div>
		</section>
	);
};

export default Tabs;
