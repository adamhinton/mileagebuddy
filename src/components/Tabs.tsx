"use client";

// ________________________________________________________________________
// This is tabs that handle internal navigation
// Has links to all internal pages
// Thanks to Flowbite, whom I shamelessly ripped this off from:
// https://flowbite.com/docs/components/tabs/
// It's sticky to the top of the page except on mobile
// Active tabs: It knows which tab the user is currently on; that tab is more visually distinct, and nothing happens when user clicks it.
// Testing: See Tabs.test.tsx
// Doesn't show Dashboard tab if user isn't logged in

// TODO Stretch tabs:
// -- Organize styling
// ________________________________________________________________________

import Link from "next/link";
import { usePathname } from "next/navigation";
import { useAppSelector } from "../redux/hooks";
import tailwindClassNames from "@/app/utils/clientUtils/styling/tailwindClassNames";

type Tab = {
	path: string;
	label: string;
};

/**Internal links to display in Tabs */
const myTabs: Tab[] = [
	{ path: "/login", label: "Login" },
	{ path: "/dashboard", label: "Dashboard" },
	{ path: "/calculator", label: "Calculator" },
	{ path: "/about", label: "About" },
];

const Tabs = () => {
	const pathname = usePathname();
	// We don't show the Dashboard tab if user isn't logged in
	const user = useAppSelector((state) => state.user.value);
	const isLoggedIn = !!user?.id;

	/**Determine whether a given tab is the page the user is currently on */
	const isActiveTab = (path: string): boolean => {
		return pathname === path;
	};

	const navStyles = tailwindClassNames.navigation;

	return (
		// Sticky to top of page, except on mobile
		<section className={navStyles.TABS_CONTAINER}>
			<div className={navStyles.TABS_INNER}>
				<ul className={navStyles.TABS_LIST}>
					{myTabs.map((tab) => {
						// Hide the Dashboard tab for non-authenticated users
						if (tab.path === "/dashboard" && !isLoggedIn) {
							return null;
						}

						// Active tab is more visually distinct, and nothing happens when user clicks it
						const isActive = isActiveTab(tab.path);
						return (
							<li key={tab.path} className={navStyles.TABS_ITEM}>
								<Link
									href={tab.path}
									aria-current={isActive ? "page" : undefined}
									className={`${navStyles.TABS_LINK} ${
										isActive 
											? navStyles.TABS_LINK_ACTIVE 
											: navStyles.TABS_LINK_INACTIVE
									}`}
									onClick={(e) => {
										if (isActive) {
											// Nothing happens because user is already on the active tab
											e.preventDefault();
										}
									}}
								>
									{tab.label}
								</Link>
							</li>
						);
					})}
				</ul>
			</div>
		</section>
	);
};

export default Tabs;
