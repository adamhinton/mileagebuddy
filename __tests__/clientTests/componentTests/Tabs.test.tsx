// ______________________________________________________________________
// This is (obviously) tests for Tabs.tsx
// These are internal navigation links
// This is rendered in the test using a dummy redux store
// TODO stretch: Test for non-authenticated users too. Not all tabs are visible to them. Just customize dummy redux store
// ______________________________________________________________________

import Tabs from "@/components/Tabs";
import { render } from "@testing-library/react";
import "@testing-library/jest-dom";
import { fireEvent } from "@testing-library/react";
// We're going to mock this
import { usePathname } from "next/navigation";
import TestReduxStore from "@/app/utils/unitTestUtils/dummyReduxStore";

jest.mock("next/navigation", () => ({
	usePathname: jest.fn(),
}));

describe("Tabs.tsx", () => {
	// Reset mocks before each test
	beforeEach(() => {
		jest.clearAllMocks();
		// Tells the component that we're on the dashboard page
		(usePathname as jest.Mock).mockReturnValue("/dashboard");
	});

	it("Renders without errors", () => {
		renderTabsWithRedux();
	});

	it("Renders all navigation tabs", () => {
		const { getByText } = renderTabsWithRedux();

		expect(getByText("Login")).toBeVisible();
		expect(getByText("Dashboard")).toBeVisible();
		expect(getByText("Calculator")).toBeVisible();
		expect(getByText("About")).toBeVisible();

		// Sanity check
		expect(() => getByText("I shouldn't exist")).toThrow();
	});

	it("Applies active styling to the current tab", () => {
		const { getByText } = renderTabsWithRedux();

		// Get the active tab element
		const activeTab = getByText("Dashboard").closest("a");

		// Check that it has the active styling classes
		expect(activeTab).toHaveClass(
			"mileage-calc-tabs-link mileage-calc-tabs-link--active"
		);

		// Should not have the hover classes that inactive tabs have
		expect(activeTab).not.toHaveClass("hover:bg-background-highlight");
	});

	it("Applies inactive styling to non-active tabs", () => {
		const { getByText } = renderTabsWithRedux();

		// Get a non-active tab element
		const inactiveTab = getByText("Login").closest("a");

		// Should not have the active styling classes
		expect(inactiveTab).not.toHaveClass("hover:bg-background-highlight");

		// Should have hover classes that active tabs don't have
		expect(inactiveTab).toHaveClass(
			"mileage-calc-tabs-link mileage-calc-tabs-link--inactive"
		);
	});

	it("Sets aria-current attribute only on active tab", () => {
		const { getByText } = renderTabsWithRedux();

		const activeTab = getByText("Dashboard").closest("a");
		const inactiveTab = getByText("Login").closest("a");

		expect(activeTab).toHaveAttribute("aria-current", "page");
		expect(inactiveTab).not.toHaveAttribute("aria-current");
	});

	it("Does not prevent default behavior when clicking inactive tabs", () => {
		const { getByText } = renderTabsWithRedux();
		const inactiveTab = getByText("Login").closest("a");

		const mockEvent = { preventDefault: jest.fn() };

		fireEvent.click(inactiveTab!, mockEvent);

		expect(mockEvent.preventDefault).not.toHaveBeenCalled();
	});
});

/**
 * Render Tabs component with a mocked Redux store
 */
const renderTabsWithRedux = () => {
	return render(
		<TestReduxStore>
			<Tabs />
		</TestReduxStore>
	);
};
