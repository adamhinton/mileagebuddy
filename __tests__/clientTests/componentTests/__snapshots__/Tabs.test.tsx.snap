// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs.tsx matches snapshot with calculator as active tab from 4.30.2025 1`] = `
<div>
  <section
    class="mileage-calc-tabs-container"
  >
    <div
      class="mileage-calc-tabs-inner"
    >
      <ul
        class="mileage-calc-tabs-list"
      >
        <li
          class="mileage-calc-tabs-item"
        >
          <a
            class="mileage-calc-tabs-link mileage-calc-tabs-link--inactive"
            href="/login"
          >
            Login
          </a>
        </li>
        <li
          class="mileage-calc-tabs-item"
        >
          <a
            class="mileage-calc-tabs-link mileage-calc-tabs-link--inactive"
            href="/dashboard"
          >
            Dashboard
          </a>
        </li>
        <li
          class="mileage-calc-tabs-item"
        >
          <a
            aria-current="page"
            class="mileage-calc-tabs-link mileage-calc-tabs-link--active"
            href="/calculator"
          >
            Calculator
          </a>
        </li>
        <li
          class="mileage-calc-tabs-item"
        >
          <a
            class="mileage-calc-tabs-link mileage-calc-tabs-link--inactive"
            href="/about"
          >
            About
          </a>
        </li>
      </ul>
    </div>
  </section>
</div>
`;

exports[`Tabs.tsx matches snapshot with dashboard as active tab from 4.30.2025 1`] = `
<div>
  <section
    class="mileage-calc-tabs-container"
  >
    <div
      class="mileage-calc-tabs-inner"
    >
      <ul
        class="mileage-calc-tabs-list"
      >
        <li
          class="mileage-calc-tabs-item"
        >
          <a
            class="mileage-calc-tabs-link mileage-calc-tabs-link--inactive"
            href="/login"
          >
            Login
          </a>
        </li>
        <li
          class="mileage-calc-tabs-item"
        >
          <a
            aria-current="page"
            class="mileage-calc-tabs-link mileage-calc-tabs-link--active"
            href="/dashboard"
          >
            Dashboard
          </a>
        </li>
        <li
          class="mileage-calc-tabs-item"
        >
          <a
            class="mileage-calc-tabs-link mileage-calc-tabs-link--inactive"
            href="/calculator"
          >
            Calculator
          </a>
        </li>
        <li
          class="mileage-calc-tabs-item"
        >
          <a
            class="mileage-calc-tabs-link mileage-calc-tabs-link--inactive"
            href="/about"
          >
            About
          </a>
        </li>
      </ul>
    </div>
  </section>
</div>
`;
